File: suites/04_compression_kernels/comparisons/block_sort.go
package main

import "fmt"

func blockSortChecksum(size int64) int64 {
	checksum := int64(0)
	for i := int64(0); i < size; i++ {
		rotChecksum := int64(0)
		for r := int64(0); r < size; r++ {
			idx := (i + r) % size
			val := (idx*7 + 13) % 256
			rotChecksum = (rotChecksum*31 + val) % 1000000007
		}
		checksum = (checksum + rotChecksum) % 1000000007
	}
	sortedCheck := int64(0)
	for j := int64(0); j < size; j++ {
		lastCol := ((j+size-1)%size*7 + 13) % 256
		sortedCheck = (sortedCheck*37 + lastCol*(j+1)) % 1000000007
	}
	return (checksum + sortedCheck) % 1000000007
}

func main() { fmt.Println(blockSortChecksum(256)) }

File: suites/04_compression_kernels/comparisons/block_sort.rs
fn block_sort_checksum(size: i64) -> i64 {
    let mut checksum: i64 = 0;
    for i in 0..size {
        let mut rot_checksum: i64 = 0;
        for r in 0..size {
            let idx = (i + r) % size;
            let val = (idx * 7 + 13) % 256;
            rot_checksum = (rot_checksum * 31 + val) % 1000000007;
        }
        checksum = (checksum + rot_checksum) % 1000000007;
    }
    let mut sorted_check: i64 = 0;
    for j in 0..size {
        let last_col = ((j + size - 1) % size * 7 + 13) % 256;
        sorted_check = (sorted_check * 37 + last_col * (j + 1)) % 1000000007;
    }
    (checksum + sorted_check) % 1000000007
}

fn main() { println!("{}", block_sort_checksum(256)); }
